### PR TITLE
feat(kit/feature): add feature flag package

### PR DESCRIPTION
**Pull request by gavincabbage**
This PR adds basic feature flag capabilities to InfluxDB. It allows users to create, consume and manipulate feature flags to exercise experimental feature before they become default behavior. The majority of code in this PR has been deployed internally for the past month and is merely open sourced by this change.

The best description of the functionality provided by this PR is in the included [`doc.go` file](https://github.com/influxdata/influxdb/pull/17851/files#diff-1a27d6fe9284b98afad5ee8b1d8a3a85):

Package feature provides feature flagging capabilities for InfluxDB servers. This document describes this package and how it is used to control experimental features in `influxd`.

Flags are configured in [`flags.yml`](https://github.com/influxdata/influxdb/pull/17851/files#diff-c1e19a99fdfbc083e8b8ec12698095f5) at the top of this repository. Running `make flags` generates Go code based on this configuration to programmatically test flag values in a given request context. [Boolean flags are the most common case, but integers, floats and strings are supported for more complicated experiments](https://github.com/influxdata/influxdb/pull/17851/files#diff-767104eff6c28be61c865f6c9fb9ba79R110-R216).

The [`Flagger`](https://github.com/influxdata/influxdb/pull/17851/files#diff-517b559f831f455e796df2ad4d267856R15-R25) interface is the crux of this package.
It computes a map of feature flag values for a given request context. The default implementation always returns the flag default configured in `flags.yml`. The override implementation allows an operator to override feature flag defaults at startup. Changing these overrides requires a restart.

In `influxd`, [a `Flagger` instance is provided](https://github.com/influxdata/influxdb/pull/17851/files#diff-95ca0f25ee9660602d6db0280db934d9R21-R24) to a [`Handler` middleware](https://github.com/influxdata/influxdb/pull/17851/files#diff-f886411bcd42ecb0c46e56a6f64582da)
configured to intercept all API requests and annotate their request context with a map of feature flags.

A flag can opt in to be exposed externally in `flags.yml`. If exposed, this flag will be included in the response from the new [`/api/v2/flags` endpoint](https://github.com/influxdata/influxdb/pull/17851/files#diff-f047f7de48d9a985ade16057d9d48739R208-R296). This allows the UI and other API clients to control their behavior according to the flag in addition to the server itself.

A concrete example to illustrate the above:

I have a feature called "My Feature" that will involve turning on new code in both the UI and the server.

First, I add an entry to `flags.yml`.

```yaml
- name: My Feature
  description: My feature is awesome
  key: myFeature
  default: false
  expose: true
  contact: My Name
```

My flag type is inferred to be boolean by my defaulf of `false` when I run `make flags` and the `feature` package now includes `func MyFeature() BoolFlag`.

I use this to control my backend code with

```go
if feature.MyFeature.Enabled(ctx) {
  // new code...
} else {
  // new code...
}
```

and the `/api/v2/flags` response provides the same information to the frontend.

```json
{
  "myFeature": false
}
```

While `false` by default, I can turn on my experimental feature by starting my server with a flag override.

```
env INFLUXD_FEATURE_FLAGS="{\"flag1\":\value1\",\"key2\":\"value2\"}" influxd
```

```
influxd --feature-flags flag1=value1,flag2=value2
```

- [x] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass
- [x] http/swagger.yml updated (if modified Go structs or API)
- [x] Documentation updated or issue created (https://github.com/influxdata/docs-v2/issues/976)

### CODE REVIEW

File: cmd/influxd/launcher/launcher.go [hunk 1]
```diff
@@ -339,6 +346,8 @@ type Launcher struct {
 	enableNewMetaStore   bool
 	newMetaStoreReadOnly bool
 
+	featureFlags string
```
Comment (Asalem1) [inline cmd/influxd/launcher/launcher.go:349]: Hey Gavin, does the list get stringified at some point or is it always a string?

File: flags.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# This file defines defines feature flags.
```
Comment (Asalem1) [inline flags.yml:1]:
```suggestion
# This file defines feature flags.
```

File: cmd/influxd/launcher/launcher.go (hunk 1 above)
Comment (gavincabbage) [reply cmd/influxd/launcher/launcher.go:349]: It's always a string and I split the list on "," myself, but it just occurred to me viper offers support for `[]string` and will do the splitting for you, so im going to refactor this to use that, which will make this bit simpler https://github.com/influxdata/influxdb/pull/17851/files#diff-5fb5403debd4552b20d5d324f7626779R38

**Comment by glinton** [comment]
Hey there @gavincabbage, i see you're into force pushing. Since you've requested reviews, please refrain from force-pushing until at least all the reviews are complete as it makes it difficult to track your recent work and wastes more time.

File: kit/feature/_codegen/main.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		name := f.Name
+		if name == "" {
+			if f.Key != "" {
+				name = f.Key
+			} else {
+				name = "anonymous"
+			}
+		}
+		// e.g. "my flag: missing key; missing default"
+		return errors.New(fmt.Sprintf("%s: %s\n", name, strings.Join(problems, "; ")))
```
[the remark this answers is not in the document]
Comment (gavincabbage) [reply kit/feature/_codegen/main.go]: Not that I can see- just an oversight. FIxed!

File: kit/cli/viper.go [hunk 4]
```diff
@@ -150,6 +150,18 @@ func BindOptions(cmd *cobra.Command, opts []Opt) {
 			}
 			mustBindPFlag(o.Flag, flagset)
 			*destP = viper.GetStringSlice(envVar)
+		case *map[string]string:
+			var d map[string]string
+			if o.Default != nil {
+				d = o.Default.(map[string]string)
+			}
+			if hasShort {
+				flagset.StringToStringVarP(destP, o.Flag, string(o.Short), d, o.Desc)
```
[the remark this answers is not in the document]
Comment (gavincabbage) [reply kit/cli/viper.go:159]: I tested with this flag both set and not and didn't run into any panics. It's just used as a default, and in `launcher.go` we don't use this map if its nil or empty anyway.

**Comment by gavincabbage** [comment]
Thanks for taking a look @desa!

File: kit/feature/doc.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+//
+// A flag can opt in to be exposed externally in `flags.yml`. If exposed,
+// this flag will be included in the response from the `/api/v2/flags`
+// endpoint. This allows the UI and other API clients to control their
+// behavior according to the flag in addition to the server itself.
+//
+// A concrete example to illustrate the above:
+//
+// I have a feature called "My Feature" that will involve turning on new code
+//// in both the UI and the server.
```
Comment (glinton) [inline kit/feature/doc.go:30]: nit:
```suggestion
// in both the UI and the server.
```

File: kit/feature/doc.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+//
+// I use this to control my backend code with
+//
+// ```go
+// if feature.MyFeature.Enabled(ctx) {
+//   // new code...
+// } else {
+//   // new code...
+// }
+//
```
Comment (glinton) [inline kit/feature/doc.go:54]: nit: no closing triple backtick. do godocs honor these now or is this just a nice add?

File: kit/feature/feature.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+		for k, v := range computed {
+			span.LogKV(k, v)
+		}
+	}
+
+	return context.WithValue(ctx, featureContextKey, computed), nil
+}
+
+// FlagsFromContext returns the map of flags attached to the context
+//by Annotate, or nil if none is found.
```
Comment (glinton) [inline kit/feature/feature.go:45]: nit here and line 55:
```suggestion
// by Annotate, or nil if none is found.
```

File: kit/feature/doc.go (hunk 6 above)
Comment (gavincabbage) [reply kit/feature/doc.go:54]: im not sure- it was intended to improve readability when reading `doc.go` directly, as opposed to the generated godocs.

**Pull request merged**